### PR TITLE
[matio] Use official CMakeLists, add features

### LINKED ISSUE #18247: [matio (x64-linux)] build failure
**Host Environment**
Windows10 20h2 / WSL2 / Ubuntu-18.04 (triplet = x64-linux)

**To Reproduce**
`./vcpkg install matio`

**Failure logs** 

> 
> Computing installation plan...
> The following packages will be built and installed:
>     matio[core]:x64-linux -> 1.5.19
> Detecting compiler hash for triplet x64-linux...
> Could not locate cached archive: /home/tim/.cache/vcpkg/archives/d5/d5a553ee8e62343cf4f0d52bdceac6bc228cc720.zip
> Starting package 1/1: matio:x64-linux
> Building package matio[core]:x64-linux...
> -- Downloading https://github.com/tbeu/matio/archive/ca56394e5672115df1981996ed12524d1d551259.tar.gz -> tbeu-matio-ca56394e5672115df1981996ed12524d1d551259.tar.gz...
> -- Extracting source /mnt/e/new_libs/scripts/vcpkg/downloads/tbeu-matio-ca56394e5672115df1981996ed12524d1d551259.tar.gz
> -- Using source at /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean
> -- Configuring x64-linux-dbg
> -- Configuring x64-linux-rel
> -- Building x64-linux-dbg
> CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:146 (message):
>     Command failed: /mnt/e/new_libs/scripts/vcpkg/downloads/tools/cmake-3.20.2-linux/cmake-3.20.2-linux-x86_64/bin/cmake --build . --config Debug --target install -- -v -j41
>     Working Directory: /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/x64-linux-dbg
>     See logs for more information:
>       /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/install-x64-linux-dbg-out.log
> 
> Call Stack (most recent call first):
>   scripts/cmake/vcpkg_build_cmake.cmake:104 (vcpkg_execute_build_process)
>   scripts/cmake/vcpkg_install_cmake.cmake:44 (vcpkg_build_cmake)
>   ports/matio/portfile.cmake:17 (vcpkg_install_cmake)
>   scripts/ports.cmake:139 (include)
> 
> 
> Error: Building package matio:x64-linux failed with: BUILD_FAILED
> Please ensure you're using the latest portfiles with `./vcpkg update`, then
> submit an issue at https://github.com/Microsoft/vcpkg/issues including:
>   Package: matio:x64-linux
>   Vcpkg version: 2021-05-05-unknownhash
> 

-Please attach any additional failure logs mentioned in the console output.

> 
> [1/13] /usr/bin/cc -DH5_BUILT_AS_STATIC_LIB -DH5_NO_DEPRECATED_SYMBOLS -DHAVE_HDF5=1 -DHAVE_ZLIB=1 -DMAT73=1 -DMATIO_HAVE_INTTYPES_H=1 -DMATIO_HAVE_STDINT_H=1 -DREPLACE_GETOPT -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/visual_studio -isystem /mnt/e/new_libs/scripts/vcpkg/installed/x64-linux/include -fPIC -g -MD -MT CMakeFiles/libmatio.dir/src/snprintf.c.o -MF CMakeFiles/libmatio.dir/src/snprintf.c.o.d -o CMakeFiles/libmatio.dir/src/snprintf.c.o -c /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c
> FAILED: CMakeFiles/libmatio.dir/src/snprintf.c.o 
> /usr/bin/cc -DH5_BUILT_AS_STATIC_LIB -DH5_NO_DEPRECATED_SYMBOLS -DHAVE_HDF5=1 -DHAVE_ZLIB=1 -DMAT73=1 -DMATIO_HAVE_INTTYPES_H=1 -DMATIO_HAVE_STDINT_H=1 -DREPLACE_GETOPT -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/visual_studio -isystem /mnt/e/new_libs/scripts/vcpkg/installed/x64-linux/include -fPIC -g -MD -MT CMakeFiles/libmatio.dir/src/snprintf.c.o -MF CMakeFiles/libmatio.dir/src/snprintf.c.o.d -o CMakeFiles/libmatio.dir/src/snprintf.c.o -c /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c: In function ‘rpl_vsnprintf’:
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:296:19: error: ‘uintptr_t’ undeclared (first use in this function); did you mean ‘__intptr_t’?
>  #define UINTPTR_T uintptr_t
>                    ^
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:747:26: note: in expansion of macro ‘UINTPTR_T’
>                          (UINTPTR_T)strvalue, 16, width,
>                           ^~~~~~~~~
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:296:19: note: each undeclared identifier is reported only once for each function it appears in
>  #define UINTPTR_T uintptr_t
>                    ^
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:747:26: note: in expansion of macro ‘UINTPTR_T’
>                          (UINTPTR_T)strvalue, 16, width,
>                           ^~~~~~~~~
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:747:36: error: expected ‘)’ before ‘strvalue’
>                          (UINTPTR_T)strvalue, 16, width,
>                                     ^~~~~~~~
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:746:21: error: too few arguments to function ‘fmtint’
>                      fmtint(str, &len, size,
>                      ^~~~~~
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/snprintf.c:404:13: note: declared here
>  static void fmtint(char *, size_t *, size_t, INTMAX_T, int, int, int, int);
>              ^~~~~~
> [2/13] /usr/bin/cc -DH5_BUILT_AS_STATIC_LIB -DH5_NO_DEPRECATED_SYMBOLS -DHAVE_HDF5=1 -DHAVE_ZLIB=1 -DMAT73=1 -DMATIO_HAVE_INTTYPES_H=1 -DMATIO_HAVE_STDINT_H=1 -DREPLACE_GETOPT -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/visual_studio -isystem /mnt/e/new_libs/scripts/vcpkg/installed/x64-linux/include -fPIC -g -MD -MT CMakeFiles/libmatio.dir/src/endian.c.o -MF CMakeFiles/libmatio.dir/src/endian.c.o.d -o CMakeFiles/libmatio.dir/src/endian.c.o -c /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/endian.c
> FAILED: CMakeFiles/libmatio.dir/src/endian.c.o 
> /usr/bin/cc -DH5_BUILT_AS_STATIC_LIB -DH5_NO_DEPRECATED_SYMBOLS -DHAVE_HDF5=1 -DHAVE_ZLIB=1 -DMAT73=1 -DMATIO_HAVE_INTTYPES_H=1 -DMATIO_HAVE_STDINT_H=1 -DREPLACE_GETOPT -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src -I/mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/visual_studio -isystem /mnt/e/new_libs/scripts/vcpkg/installed/x64-linux/include -fPIC -g -MD -MT CMakeFiles/libmatio.dir/src/endian.c.o -MF CMakeFiles/libmatio.dir/src/endian.c.o.d -o CMakeFiles/libmatio.dir/src/endian.c.o -c /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/endian.c
> In file included from /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/visual_studio/matio_pubconf.h:127:0,
>                  from /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/matio.h:34,
>                  from /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/matio_private.h:31,
>                  from /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/src/endian.c:30:
> /mnt/e/new_libs/scripts/vcpkg/buildtrees/matio/src/4d1d551259-c433863f19.clean/visual_studio/stdint_msvc.h:34:2: error: #error "Use this header only with Microsoft Visual C++ compilers!"
>  #error "Use this header only with Microsoft Visual C++ compilers!"
>   ^~~~~
> LOG CONTINUES with same error for 2000 more lines

**Additional context**
Hello, it appears that the matio cmake build is totally broken on linux. During compilation, a header file directly complains "#error "Use this header only with Microsoft Visual C++ compilers!"

I would be willing to help fix this but I am new to cmake and vcpkg and I dont really know what I am doing. I did a quick look into the cmake code that vcpkg has, and the cmake code that comes with matio, and it looks like the provided build system from matio is much more featured and 'working' than the one that comes with vcpkg, which looks like someone unfamilliar with matio just wrote a quick cmake file that only works on one platform. The cmake files provided by the maintainers of matio do many specific checks for header files and set flags accordingly, whereas the VCPKG port just hardcodes a few of the values for windows platforms

1) is this intended? to entirely replace the cmake build system with a new one handwritten by the vcpkg maintainers? I understand we might need to replace checks for certain libraries so that we always use the vcpkg version of any dependencies (in this case, hdf5, zlib) but the amount of reduction that has been done here seems excessive
2) would the correct thing be to copy the matio cmake files into the vcpkg port and then slightly modify them? I have tried reading various documents on how to create ports but I feel like i haven't found the 'core" tutorial that addresses basic things like this. It seems like it would have been a lot easier to use the existing CMAKE files and just modify the 'thirdparty.cmake' file that exclusively deals with finding HDF5 and zlib. 

I know for a fact that i can build this version of matio on my platform using its own CMAKE files because I have done so. the only issue is that it is not handled properly by vcpkg

### PR DESCRIPTION
**Pull request by JackBoosY**
Use official cmake system, add features.

Fixes #18247.

Already tested all features successfully on `x86-windows` and `x64-windows-static`.

### CODE REVIEW

**Comment by JackBoosY** [comment]
@timmeh87  Can you please test this PR?

Thanks.

File: versions/m-/matio.json [hunk 1]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "d027de2222790b013d47eb9e0dd1fd8004087bec",
```
Comment (JackBoosY) [inline versions/m-/matio.json:4]:
```suggestion
      "git-tree": "b49de24bd5462214863582762ae16850e4ab7a4d",
```

**Comment by timmeh87** [comment]
Hello, I tested this PR on my local x64-linux (WSL2 ubuntu18.04) and it works for me. Thanks!

File: ports/matio/portfile.cmake [hunk 2]
```diff
@@ -4,18 +4,34 @@ vcpkg_from_github(
...
-file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/matio RENAME copyright)
+vcpkg_cmake_install()
 
 vcpkg_copy_pdbs()
+
+vcpkg_copy_tools(TOOL_NAMES matdump AUTO_CLEAN)
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+
+file(INSTALL "${SOURCE_PATH}/COPYING" DESTINATION "${CURRENT_PACKAGES_DIR}/share/matio" RENAME copyright)
```
Comment (JackBoosY) [inline ports/matio/portfile.cmake:37]:
```suggestion
file(INSTALL "${SOURCE_PATH}/COPYING" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```

File: ports/matio/portfile.cmake [hunk 3]
```diff
@@ -4,18 +4,34 @@ vcpkg_from_github(
...
+string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" BUILD_SHARED)
 
-vcpkg_configure_cmake(
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+    hdf5            MATIO_WITH_HDF5
+    zlib            MATIO_WITH_ZLIB
+    extended-sparse MATIO_EXTENDED_SPARSE
+    mat73           MATIO_MAT73
+    pic             MATIO_PIC
```
Comment (JackBoosY) [inline ports/matio/portfile.cmake:18]:
```suggestion
    FEATURES
        hdf5            MATIO_WITH_HDF5
        zlib            MATIO_WITH_ZLIB
        extended-sparse MATIO_EXTENDED_SPARSE
        mat73           MATIO_MAT73
        pic             MATIO_PIC
```

**Review by NancyLi1013 — APPROVED** [review]

File: versions/m-/matio.json [hunk 4]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "b49de24bd5462214863582762ae16850e4ab7a4d",
```
Comment (JackBoosY) [inline versions/m-/matio.json:4]:
```suggestion
      "git-tree": "c11c8ba1ca5c7f0286165d497088779ec0df756a",
```

**Comment by strega-nil-ms** [comment]
LGTM, thanks @JackBoosY

**Pull request merged**